File: packages/dotnet/CatalystSharp/Clients/EpicleseClient.cs
using CatalystSharp.Http;
using CatalystSharp.Models;
using CatalystSharp.Models.Requests;

namespace CatalystSharp.Clients;

public class EpicleseClient
{
    private readonly ICatalystHttpClient _httpClient;

    internal EpicleseClient(ICatalystHttpClient httpClient)
    {
        _httpClient = httpClient;
    }

    public async Task<EpicleseAuthorsResult> GetAuthorsAsync(string? query = null, string? platform = null, CancellationToken cancellationToken = default)
    {
        return await _httpClient.GetAsync<EpicleseAuthorsResult>("/epiclese/v1/authors", new Dictionary<string, string?>
        {
            ["q"] = query,
            ["platform"] = platform
        }, cancellationToken);
    }

    public async Task<EpicleseAuthor> CreateAuthorAsync(CatalystCreateMetadataAuthorRequest request, CancellationToken cancellationToken = default)
    {
        var response = await _httpClient.PostAsync<EpicleseAuthorWrapper>("/epiclese/v1/authors", request, cancellationToken);
        return response.Author;
    }

    public async Task<IReadOnlyList<EpiclesePlatform>> GetPlatformsAsync(CancellationToken cancellationToken = default)
    {
        var response = await _httpClient.GetAsync<EpiclesePlatformsWrapper>("/epiclese/v1/platforms", cancellationToken);
        return response.Platforms;
    }

    public async Task<EpiclesePlatform> GetPlatformAsync(string id, CancellationToken cancellationToken = default)
    {
        var response = await _httpClient.GetAsync<EpiclesePlatformWrapper>($"/epiclese/v1/platforms/{id}", cancellationToken);
        return response.Platform;
    }

    public async Task<IReadOnlyDictionary<string, EpicleseStatusMetadataTag>> GetStatusMetadataAsync(string statusId, CancellationToken cancellationToken = default)
    {
        return await _httpClient.GetAsync<IReadOnlyDictionary<string, EpicleseStatusMetadataTag>>($"/epiclese/v1/tag/by/status/{statusId}", cancellationToken);
    }

    public async Task<IReadOnlyDictionary<string, EpicleseStatusMetadataTag>> CreateStatusMetadataAsync(string statusId, IReadOnlyList<CatalystCreateStatusMetadataItem> request, CancellationToken cancellationToken = default)
    {
        return await _httpClient.PostAsync<IReadOnlyDictionary<string, EpicleseStatusMetadataTag>>($"/epiclese/v1/tag/by/status/{statusId}", request, cancellationToken);
    }

    public async Task<EpicleseWorldsResult> GetWorldsAsync(string? query = null, string? platform = null, int? offset = null, CancellationToken cancellationToken = default)
    {
        return await _httpClient.GetAsync<EpicleseWorldsResult>("/epiclese/v1/worlds", new Dictionary<string, string?>
        {
            ["q"] = query,
            ["platform"] = platform,
            ["offset"] = offset?.ToString()
        }, cancellationToken);
    }

    public async Task<EpicleseWorld> CreateWorldAsync(CatalystCreateMetadataWorldRequest request, CancellationToken cancellationToken = default)
    {
        var response = await _httpClient.PostAsync<EpicleseWorldWrapper>("/epiclese/v1/worlds", request, cancellationToken);
        return response.World!;
    }

    public async Task<EpicleseWorld?> ResolveWorldAsync(string platform, string name, CancellationToken cancellationToken = default)
    {
        var response = await _httpClient.GetAsync<EpicleseWorldWrapper>("/epiclese/v1/worlds/resolve", new Dictionary<string, string?>
        {
            ["platform"] = platform,
            ["name"] = name
        }, cancellationToken);
        return response.World;
    }
}

File: packages/kotlin/src/main/kotlin/com/natsuneko/catalyst/clients/EpicleseClient.kt
// Licensed under the MIT License
//
// Copyright (c) 2025 Natsune Mochizuki

package com.natsuneko.catalyst.clients

import com.natsuneko.catalyst.http.CatalystHttpClient
import com.natsuneko.catalyst.models.*

/**
 * Client for Epiclese (World/Author metadata) API endpoints
 */
class EpicleseClient internal constructor(
    private val httpClient: CatalystHttpClient
) {
    suspend fun getAuthors(query: String? = null, platform: String? = null): EpicleseAuthorsResult = httpClient.get(
        "/epiclese/v1/authors",
        mapOf("q" to query, "platform" to platform)
    )

    suspend fun createAuthor(
        platformIdentifier: String,
        platform: String,
        name: String,
        id: String? = null
    ): EpicleseAuthor = httpClient.postWithResult<EpicleseAuthorWrapper>(
        "/epiclese/v1/authors",
        mapOf(
            "platformIdentifier" to platformIdentifier,
            "platform" to platform,
            "name" to name,
            "id" to id
        ).filterValues { it != null }
    ).author

    suspend fun getPlatforms(): List<EpiclesePlatform> =
        httpClient.get<EpiclesePlatformsWrapper>("/epiclese/v1/platforms").platforms

    suspend fun getPlatform(id: String): EpiclesePlatform =
        httpClient.get<EpiclesePlatformWrapper>("/epiclese/v1/platforms/$id").platform

    suspend fun getStatusMetadata(statusId: String): Map<String, EpicleseStatusMetadataTag> =
        httpClient.get("/epiclese/v1/tag/by/status/$statusId")

    suspend fun createStatusMetadata(
        statusId: String,
        tags: List<Map<String, Any?>>
    ): Map<String, EpicleseStatusMetadataTag> = httpClient.postWithResult(
        "/epiclese/v1/tag/by/status/$statusId",
        tags
    )

    suspend fun getWorlds(
        query: String? = null,
        platform: String? = null,
        offset: Int? = null
    ): List<EpicleseWorld> = httpClient.get<EpicleseWorldsResult>(
        "/epiclese/v1/worlds",
        mapOf("q" to query, "platform" to platform, "offset" to offset?.toString())
    ).items

    suspend fun createWorld(
        platformIdentifier: String,
        platform: String,
        name: String,
        author: Map<String, Any?>? = null,
        id: String? = null
    ): EpicleseWorld? = httpClient.postWithResult<EpicleseWorldWrapper>(
        "/epiclese/v1/worlds",
        mapOf(
            "platformIdentifier" to platformIdentifier,
            "platform" to platform,
            "name" to name,
            "author" to author,
            "id" to id
        ).filterValues { it != null }
    ).world

    suspend fun resolveWorld(platform: String, name: String): EpicleseWorld? = httpClient.get<EpicleseWorldWrapper>(
        "/epiclese/v1/worlds/resolve",
        mapOf("platform" to platform, "name" to name)
    ).world
}
